js/src/vm/PortableBaselineInterpret.cpp

```diff
@@ -4402,13 +4402,18 @@ static PBIResult PortableBaselineInterpret(JSContext* cx_, State& state,
 
     CASE(CheckReturn) {
       Value thisval = POP().asValue();
-      if (frame->returnValue().isObject()) {
-        PUSH(StackVal(frame->returnValue()));
-      } else if (!ret->isUndefined()) {
+      // inlined version of frame->checkReturn(thisval, result) (js/src/vm/Stack.cpp)
+      // except we call PUSH_EXIT_FRAME before any function call to report an error
+      // and if checkReturn would return false then we `goto error`
+      // and otherwise we PUSH(StackVal(result))
+      HandleValue retVal = frame->returnValue();
+      if (retVal.isObject()) {
+        PUSH(StackVal(retVal));
+      } else if (!retVal.isUndefined()) {
         PUSH_EXIT_FRAME();
-        state.value0 = frame->returnValue();
-        ReportValueError(cx, JSMSG_BAD_DERIVED_RETURN, JSDVG_IGNORE_STACK,
-                         state.value0, nullptr);
+        state.value0 = retVal; // is this necessary?
```

Comment:
Shouldn't be needed -- the `Handle` type denotes that the value is already rooted (a handle is a pointer to the presumed-rooted storage), and in fact the `BaselineFrame` does root its return-value slot when traced; so passing `retVal` below without storing it in the `state` here is perfectly fine.